### PR TITLE
Prepare React 16 Update

### PR DESCRIPTION
**Pull request by chapati23**
"Just" wanted to try make this work with `auth0-lock` but went down a rabbit hole of updating all the things. 🕳🐇 — Hope some of this stuff is useful, open to change some things but will switch to moving my fork for now

### Done
* [x] Updated all npm packages to latest possible versions
* [x] Updated Flow and fixed all flow warnings/errors
* [x] Updated Eslint and fixed all lint warnings/errors (except for 1)
* [x] Update storybook setup
* [x] Created “yarn start” task to bring up storybook
* [x] Created “yarn lint” task to run eslint over all *.js files
* [x] Include polyfills to prepare tests for React 16

### Open for another MR
- [ ] Refactor or remove `jest-redux-snap`, seems to be the only thing blocking the React 16 migration
- [ ] Update to React 16
- [ ] Migrate to use the latest version of [react-transition-group](https://github.com/reactjs/react-transition-group)

### CODE REVIEW

**Comment by faceyspacey** [comment]
woah! ...I'm open to whatever you want to do for the tests. A lot of what `jest-redux-snap` solves is no longer a problem. And to begin with it was a little unnecessary for this package--I only used it because it was easy for me. With the new React testing apis, we likely can just pass the new `props` and update the component to test what happens, skipping the redux craziness. 

Keep up the good work. It sounds like you got another pass planned. I'll go through it in depth then, but I trust what you're doing.

**Comment by chapati23** [comment]
I took a detour and actually refactored one of my projects to use `react-transition-group` directly. With [callbacks now enabled](https://reactcommunity.org/react-transition-group/#CSSTransition), is there still a use-case that you can't solve with `react-transition-group`?

**Comment by faceyspacey** [comment]
hmm. quick comparison:

**transition-group PROs:**
- delays
- global options that are automatically passed down to transitions
- state-level delay + duration options (eg: `appearDelay`)
- no confusion about when to use Transition vs TransitionGroup. this one always requires a group, and handles multiple transitions the same as a single transition, which brings us to the next point:
- automatic handling of simple  single transition groups: https://github.com/faceyspacey/transition-group/blob/master/src/TransitionGroup.js#L83 -- i.e. this stuff is the basis for a consistent API where you don't have to remember the difference between TG and T and multiple Ts vs a single T. It's always the same. TG > T(s)
- group `onFull` + `onEmpty` callbacks

**new RTG PROs/doubts:**
- node.addEventListener thing (as i recall, its purpose, `transitionend` is problematic in many browsers), so this is more of a hopeful feature, than a practical one
- the `in` option seems like automation to not have to do `{someBool && <Transition...`. You're gonna have to do the latter anyway with dynamic lists--probably better to keep the API surface more minimal and not introduce this "automagic," but it's useful for static transition components
- `childFactory` allows you to change states of animating/transitioning elements. this is useful, but at the same time, potentially dangerous, as i dont care how good they say Fiber is, the browser sucks, and animations coinciding with render updates always produces jank. so it's probably best to avoid this.
- probably minorly better performing
- probably better implementation of css class applying + callbacks. the way i set it up doesn't provide a guarantee that the callbacks are called exactly one tick after a transition ends. I'm less concerned about this than purists might be. It just doesn't matter 99.999% of the time. I'm not sure how they solved it--perhaps they didn't, and they just timed it similarly like I did, or rather like you're supposed to in userland by setting matching animation times.
- enable/disable transitions?? ...probably will get very little use, as it defeats the purpose, but of course there are times for such things. we probably can accomplish it via current features, or just add the feature. 
- function as children. this could be interesting.
- the mountOnEnter stuff seems like noise. noise to address the `in` option. I'm not exactly sure of what behavior this will produce, but it seems like a substitute for the `appear` vs `enter` stuff, for which we already have handling.

Overall, i dont like the API. it's not as consistent, and easy to remember, not as granularly + globally customizeable, and solves very few additional problems. Pretty much none besides `childFactory`, but please correct me if I'm missing some magic. Also, the current transition-group works in React 16. The RFR boilerplate + demo have been updated and it works. Perhaps once fiber is out, it will have additional challenges--is Fiber out yet?

So basically, they now both support callbacks + css. It's annoying, because it leaves me wondering if we're missing out on perf benefits if we dont upgrade to be backed by the latest RTG.

Anyway, another day in Reactlandia...

**Comment by chapati23** [comment]
yeah, i'm conflicted. otoh i really like all the shit you're doing, otoh using `transition-group` vs RTG most likely means being confined to older versions and relying mainly on you for updates. and i already got the latest RTG working exactly the same as your `redux-first-router-demo` example.

did you try convince RTG to add some of the functionality you're missing?

**Comment by faceyspacey** [comment]
Older versions, updates lol, convince them? Rtg was almost thrown in the garbage by them. I'm surprised they updated it. My prediction is there will be few updates. At least not for like 2 years. They don't care about that component. At least they didn't before, and finally came to their senses just recently and realized/remembered the importance of rtg. 

That said, a good version of this doesn't need any updates, it's supposed to be a straightforward component. The only updates should just be dep updates.

I could convince them, but it would likely fall on deaf ears. Also look in their code, I wanna say it's not that hard, but it is a bit complex. However it's not a lot of code, and it's not magic. I'd be more enticed to rebuild it as an exercise. This precise code actually is some of the most important code in the React game. It teaches how what a component renders can be very different than the props/children it receives. 

At the end of the day I doubt in terms of perf it makes any difference if this package upgrades the rtg backing it. I'll look into their latest code--I bet the core state inversion of control mechanism is exactly the same. Just a different API, and a few minor new features. The API they chose is awkward and unnecessarily complex, so my hunch is that a junior developer stepped up to the plate to take it over. Or some open source developer that has minimal *product* experience. It's not a good experience for product developers. This one is, and here's why:

*tg > t(s). Same options in both components. Automatic handling of one component vs multiple. Never use t without tg. Easy to remember.*

When I get around to updating this, I'm gonna seriously consider just copy/pasting the core inversion of control mechanism and dropping the RTG dep. I may never find the time, but I'll at least see if much has changed, and if much hasn't, the dep may in fact not need to change. After all React 16 and fiber isn't supposed to be a huge departure from before. Probably 95% of code still works perfectly, which clearly encompasses this code. Any perf gains are likely because the new RTG itself tried to be faster.

**Comment by faceyspacey** [comment]
u know what would be even cooler: a glamorous or styled-components specific version of this where you don't have to specify the timing info twice (in js + css). u just specify the css for each class in javascript, and then the delays and durations are added to the css automatically for you. 

this would also eliminate options like prefixes, and generally having to worry about classNames at all. you just specify the css for each phase/state. 

i got some things planned for glamorous (precompiling static sheets + pre-evaluating js/css at build time, so you can use dynamic js in what becomes your static css; for all css that doesn't take props; css that depends on props is handled the way glamorous already does). To me, that's the final nail in the coffin of traditional css and completes the vision, at which point I don't even want to use css classes anymore anyway! So then might be a good time to revisit this.

**Comment by SeabassNg** [comment]
So, what's the update for this PR?

**Comment by faceyspacey** [comment]
It needs to be finished

**Comment by SeabassNg** [comment]
@faceyspacey I might try my hand on this. What exactly do we need left here?

**Comment by SeabassNg** [comment]
@chapati23

**Comment by chapati23** [comment]
@SeabassNg sry man, i kinda abandoned this PR after deciding to just use the latest `react-transition-group` because i didn't need any of the advanced features of `transition-group`. i don't really remember what's left, but the 3 open tasks in the PR description should be an indication
